### PR TITLE
JBPM-7798 After login, loading screen localized message is messed up

### PR DESCRIPTION
**Pull request by bxf12315**
@cristianonicolai Please review it.

### CODE REVIEW

**Comment by kiereleaseuser2** [comment]
Build finished. 2761 tests run, 5 skipped, 0 failed.

**Comment by kiereleaseuser2** [comment]
Build finished. 2761 tests run, 5 skipped, 0 failed.

File: errai-bus/src/main/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilter.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 import javax.servlet.http.HttpServletResponse;
 import javax.servlet.http.HttpSession;
 
+import org.jboss.errai.common.server.FilterCacheUtil.CharResponseWrapper;
```
Comment (cristianonicolai) [inline errai-bus/src/main/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilter.java]: I believe you don't need this import any longer.

Comment (bxf12315) [reply errai-bus/src/main/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilter.java]: new code need it

**Comment by bxf12315** [comment]
@cristianonicolai  I try to add unit test for it.
If you open code line of CSRFTokenFilterTest.java (when(resp.getOutputStream()).thenReturn(getServletOutputStream())) and use response.getOutputStream().write(bytes);  instead of response.getWriter().print(output);
You can got error message like this:
ERROR] Tests run: 2, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.739 s <<< FAILURE! - in org.jboss.errai.bus.server.servlet.CSRFTokenFilterTest
[ERROR] testJapanese(org.jboss.errai.bus.server.servlet.CSRFTokenFilterTest)  Time elapsed: 0.163 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[￣ﾁﾗ￣ﾁﾰ￣ﾂﾉ￣ﾁﾏ￣ﾁﾊ￥ﾾﾅ￣ﾁﾡ￣ﾁﾏ￣ﾁﾠ￣ﾁﾕ￣ﾁﾄ]> but was:<[しばらくお待ちください]>
	at org.jboss.errai.bus.server.servlet.CSRFTokenFilterTest.testJapanese(CSRFTokenFilterTest.java:99)

[ERROR] testChinese(org.jboss.errai.bus.server.servlet.CSRFTokenFilterTest)  Time elapsed: 0.005 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[￨ﾯﾷ￧ﾨﾍ￥ﾀﾙ]> but was:<[请稍候]>
	at org.jboss.errai.bus.server.servlet.CSRFTokenFilterTest.testChinese(CSRFTokenFilterTest.java:87)

**Comment by kiereleaseuser2** [comment]
Build finished. 2765 tests run, 5 skipped, 0 failed.

File: errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        when(req.getSession(false)).thenReturn(session);
+        when(req.getMethod()).thenReturn("GET");
+        when(resp.getWriter()).thenReturn(new PrintWriter(charWriter));
+        when(resp.getContentType()).thenReturn("text/html");
+        when(resp.getOutputStream()).thenReturn(getServletOutputStream());
+    }
+
+    @Test
+    public void testChinese() throws IOException, ServletException {
+        CSRFTokenFilter csrfTokenFilter = getFilterForChinese();
```
Comment (cristianonicolai) [inline errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java]:
```suggestion
                doAnswer(ctx -> {
            final InputStream stream = getClass().getResourceAsStream("/test-html/sample-Japanese.html");
            final String text = IOUtils.toString(stream);
            ((HttpServletResponse) ctx.getArguments()[1]).getWriter().print(text);
            return null;
        }).when(chain).doFilter(any(),
                                any());

        CSRFTokenFilter csrfTokenFilter = new CSRFTokenFilter();
```

File: errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+        final Document document = Jsoup.parse(charWriter.toString());
+        Elements es = document.select("div.center-block");
+
+        assertEquals(es.get(1).text(), japanese_loadingPleaseWait);
+        assertEquals(es.get(2).text(), japanese_loadingApplication);
+    }
+
+    @Test
+    public void testErrorChinese() throws IOException, ServletException {
```
Comment (cristianonicolai) [inline errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java]: Please remove test from old behavior.

File: errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+        final Document document = Jsoup.parse(charWriter.toString());
+        Elements es = document.select("div.center-block");
+
+        assertEquals(es.get(1).text(), chinese_loadingPleaseWait_error);
+        assertEquals(es.get(2).text(), chinese_loadingApplication_error);
+    }
+
+    @Test
+    public void testErrorJapanese() throws IOException, ServletException {
```
Comment (cristianonicolai) [inline errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java]: Please remove test from old behavior.

**Comment by bxf12315** [comment]
@cristianonicolai According to your comment, I updated the PR, Please review it.

**Comment by cristianonicolai** [comment]
looks good to me.

**Comment by cristianonicolai** [comment]
@tiagobento please take a look.

**Comment by kiereleaseuser2** [comment]
Build finished. 2763 tests run, 5 skipped, 0 failed.

**Comment by tiagobento** [comment]
@cristianonicolai @bxf12315 Hi! So from what I understand, this PR solves JBPM-7798 and #367 adds a feature to Errai, am I correct?

If so, I think this PR is good to merge, provided it's tested and the Workbench has been run to confirm it works. Did someone from QE take a look?

About #367, if it's not needed to solve JBPM-7798, I'd ask for you to change the PR's title and create a JIRA with a description of the future proposal so I can better understand the intent.

Thanks!

**Comment by bxf12315** [comment]
@tiagobento You are right. The PR solves issue JBPM-7789. It is my first way to fix the issue.
https://github.com/errai/errai/pull/367, is my second way and I think it is better way.
Because of  , org.jboss.errai.security.server.servlet.UserHostPageFilter, org.jboss.errai.bus.server.servlet.CSRFTokenFilter,org.uberfire.server.locale.GWTLocaleHeaderFilter all inject some script in target page. So I try to use one Filter to do this function like InjectScriptFilter instead of it.
The reason for causing JBPM-7789 is getOutputStream().write(bytes) of CharResponseWrapper, can not support Chinese and Japanese. UserHostPageFilter, CSRFTokenFilter, GWTLocaleHeaderFilter wrapped  CharResponseWrapper 3 times, In CSRFTokenFilter and GWTLocaleHeaderFilter, response is instance of CharResponseWrapper, so the issue will occurred. In the same time , I think if we add new Inject script filter (like CSRFTokenFilter...) or change order of the filter chain, we also will met same problem.
Thanks

**Comment by cristianonicolai** [comment]
@bxf12315 the main issue here is using `response.getOutputStream().write(bytes);` instead of `response.getWriter().print(output);`. If you apply the same fix in here https://github.com/errai/errai/blob/master/errai-security/errai-security-server/src/main/java/org/jboss/errai/security/server/servlet/UserHostPageFilter.java#L103 then the other of the filters won't be an issue. @tiagobento wdyt?

**Comment by tiagobento** [comment]
@cristianonicolai @bxf12315 I agree with Cristiano. The changes made on this PR seem a lot less intrusive and are simple enough to be replicated where needed. I think we should favor this PR over #367.

Please tell me if you need me to merge it. 

@cristianonicolai I'm not sure if someone from QE had the chance to look at this feature's PRs. If so, can you mention them here so they can be aware that it has impact on Errai too?

Thanks!

**Comment by cristianonicolai** [comment]
@tiagobento there is no one assigned from QE to help with this atm. @jhrcek @MarianMacik could you help review this or get someone involved, thanks.

File: errai-bus/src/main/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilter.java [hunk 5]
```diff
@@ -74,19 +73,24 @@ public void doFilter(final ServletRequest request, final ServletResponse respons
...
       if (session != null && responseContentType != null && responseContentType.toLowerCase().startsWith("text/html")) {
+
         CSRFTokenCheck.INSTANCE.prepareSession(session, log);
         final Document document = Jsoup.parse(responseWrapper.toString());
-        document.head().prepend("<script>var " + ERRAI_CSRF_TOKEN_VAR + " = '" + CSRFTokenCheck.getToken(session) + "';</script>");
-        bytes = document.html().getBytes("UTF-8");
+        String injectedScript = "<script>var " + ERRAI_CSRF_TOKEN_VAR + " = '" + CSRFTokenCheck.getToken(session) + "';</script>";
+        document.head().prepend(injectedScript);
+        output = document.html();
+        bytes = output.getBytes("UTF-8");
```
Comment (jhrcek) [inline errai-bus/src/main/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilter.java]: Nitpick: The line `bytes = output.getBytes("UTF-8")` is identical in both if and else branches. It should be moved after the if-else.

File: errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        final Document document = Jsoup.parse(charWriter.toString());
+        Elements es = document.select("div.center-block");
+
+        assertEquals(chinese_loadingPleaseWait, es.get(1).text());
+        assertEquals(chinese_loadingApplication, es.get(2).text());
+    }
+
+    @Test
+    public void testJapanese() throws IOException, ServletException {
+        CSRFTokenFilter csrfTokenFilter = new CSRFTokenFilter();
```
Comment (jhrcek) [inline errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java]: These test methods seem identical, differing only in parameters. I'd refactor the common code into helper method and also would give more informative name to the method. Like so:

```java
void csfrTokenFilterShouldntAffectJapaneseCharacters() {
    csfrTokenFilterTest("/test-html/sample-Japanese.html", japanese_loadingPleaseWait, japanese_loadingApplication);
}
```

**Comment by jhrcek** [comment]
Verified the fix locally, it works fine.

File: errai-bus/src/main/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilter.java (hunk 5 above)
Comment (bxf12315) [reply errai-bus/src/main/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilter.java]: ok

File: errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java (hunk 6 above)
Comment (bxf12315) [reply errai-bus/src/test/java/org/jboss/errai/bus/server/servlet/CSRFTokenFilterTest.java]: ok

**Comment by bxf12315** [comment]
Done

**Comment by tiagobento** [comment]
Thanks @jhrcek. Please tell me when it's ready to merge @cristianonicolai @bxf12315.

**Comment by kiereleaseuser2** [comment]
Build finished. 2763 tests run, 5 skipped, 0 failed.

**Pull request merged**